src/combinatorics/simple_graph/connectivity.lean

```diff
@@ -1502,23 +1519,108 @@ lemma preconnected.subsingleton_connected_component (h : G.preconnected) :
+{ to_fun := connected_component.map φ.to_hom,
+  inv_fun := connected_component.map φ.symm.to_hom,
+  left_inv := λ C, connected_component.ind
+    (λ v, congr_arg (G.connected_component_mk) (equiv.left_inv φ.to_equiv v)) C,
+  right_inv := λ C, connected_component.ind
+    (λ v, congr_arg (G'.connected_component_mk) (equiv.right_inv φ.to_equiv v)) C }
+
+/-- The set of vertices in a connected component of a graph. -/
+def connected_component.supp (C : G.connected_component) :=
+  { v | G.connected_component_mk v = C }
+
+@[ext] lemma connected_component.supp_injective :
+  function.injective (connected_component.supp : G.connected_component → set V) :=
+begin
+  refine connected_component.ind₂ _,
+  intros v w h,
+  simp only [set.ext_iff, connected_component.eq, set.mem_set_of_eq, connected_component.supp] at
+     h ⊢,
+  exact ((h v).mp (reachable.refl _)),
+end
+
+@[simp]
+lemma connected_component.supp_inj {C D : G.connected_component} : C.supp = D.supp ↔ C = D :=
+connected_component.supp_injective.eq_iff
+
+instance : set_like G.connected_component V :=
+{ coe := connected_component.supp,
+  coe_injective' := connected_component.supp_injective, }
+
+@[simp] lemma connected_component.mem_supp_iff (C : G.connected_component) (v : V) :
+  v ∈ C.supp ↔ G.connected_component_mk v = C := iff.rfl
+
+lemma connected_component_mk_mem (G : simple_graph V) {v : V} :
+  v ∈ G.connected_component_mk v := by { exact rfl, }
+
+/--
+The equivalence between connected components, induced by an isomorphism of graphs,
+itself defines an equivalence on the supports of each connected component.
+-/
+def connected_component.iso_equiv_supp (φ : G ≃g G') (C : G.connected_component) :
+  C.supp ≃ (connected_component.iso φ C).supp :=
+{ to_fun := λ v, ⟨φ.to_fun v.val, connected_component.iso_image_comp_eq_map_iff_eq_comp.mpr v.prop⟩,
+  inv_fun := λ v', ⟨φ.inv_fun v', connected_component.iso_inv_image_comp_eq_iff_eq_map.mpr v'.prop⟩,
+  left_inv := λ v, subtype.ext_val (φ.to_equiv.left_inv ↑v),
+  right_inv := λ v, subtype.ext_val (φ.to_equiv.right_inv ↑v), }
+
+protected lemma connected_component.connected :
+  ∀ (C : G.connected_component), (G.induce {v : V | G.connected_component_mk v = C}).connected :=
+begin
+  refine connected_component.ind (λ v, _),
+  refine (connected_iff _).mpr ⟨_, ⟨⟨v, rfl⟩⟩⟩,
+  let Gr := G.reachable,
+  rw set.ext (λ (v' : V), @connected_component.eq V G v' v),
+  let GCr := (G.induce {t : V | Gr t v}).reachable,
+  have : ∀ {u w} (p : G.walk u w) (hu : Gr u v) (hw : Gr w v), GCr ⟨u,hu⟩ ⟨w,hw⟩, by
+  { rintro u w p,
+    induction p with _ _ _ _ a q ih,
+    { exact λ _ _, reachable.refl _, },
+    { refine λ hu hw, reachable.trans (adj.reachable _) (ih (a.reachable.symm.trans hu) hw),
+      exact a, }, },
+  rintro ⟨u,hu⟩ ⟨w,hw⟩,
+  simp only [connected_component.eq, set.mem_set_of_eq] at hu hw,
+  exact this (reachable.trans hu $ reachable.symm hw).some hu hw,
+end
```

Comment:
I don't think this lemma is ready for the library. To do it right, we should have supporting theory for relating walks in `G.induce s` to walks whose support is contained in `s`.

I started writing
```

def induce_lift (s : set V) : Π {v w : V} (p : G.walk v w) (h : ∀ u, u ∈ p.support → u ∈ s),
  (G.induce s).walk ⟨v, h v p.start_mem_support⟩ ⟨w, h w p.end_mem_support⟩
| _ _ walk.nil _ := walk.nil
| _ _ (walk.cons' u v w ha p) h :=
  let p' := induce_lift p (λ x hx, h x (by simp [hx]))
  in walk.cons (by exact ha) p'
```
but fleshing everything out would take some time. Would you be ok with removing the lemma for now?

Reply:
Thanks for the review! I agree having a more general theory for mapping walks back and forth would be better (this is how we handled it before), but at the same time the lemma works as-is and can be made to use this API when it gets there…
Since it's needed to make progress on Freudenthal-Hopf, I'd rather have it there, but I understand if you don't want to introduce "half-baked" results in `simple_graph.*`. Your call in any case :)

Reply:
I get that it works as-is, but at least for now it seems somewhat out of place from the rest of the PR.

I'm dusting off some things I have in a branch to make use of what you have here. I've said it before, but I suspect a number of things are cleaner through subgraphs rather than induced graphs (or at least, if you want to work with induced graphs, they can follow from facts about subgraphs).

In the meantime, would you be ok keeping this lemma in your Freudenthal-Hopf project?

Reply:
Yeah, works for me :)